luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,217 @@
+local spGetUnitEstimatedPath = Spring.GetUnitEstimatedPath
+local mathMin = math.min
+local mathMax = math.max
+
+--tables
+local smartUnits = {}
+local unitDefData = {}
+
+function gadget:Initialize()
+	local units = Spring.GetAllUnits()
+	local spGetUnitDefID = Spring.GetUnitDefID
+	for i = 1, #units do
+		gadget:UnitCreated(units[i], spGetUnitDefID(units[i]))
+	end
+end
+
+for unitDefID, def in ipairs(UnitDefs) do
+	if def.weapons then
+		local weapons = def.weapons
+		for weaponNumber, weaponData in pairs(weapons) do
+			local weaponDefID = weapons[weaponNumber].weaponDef
+			if WeaponDefs[weaponDefID] and WeaponDefs[weaponDefID].customParams then
+				if WeaponDefs[weaponDefID].customParams.smart_priority then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].priorityWeapon = weaponNumber
+					unitDefData[unitDefID].failedToFireFrameThreshold = WeaponDefs[weaponDefID].customParams.smart_error_frames or
+						mathMax(WeaponDefs[weaponDefID].reload * failedToFireMultiplier, minimumFailedToFireFrames)
+					if def.speed and def.speed ~= 0 then
+						unitDefData[unitDefID].canMove = true
+					end
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_backup then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].backupWeapon = weaponNumber
+				end
+				if WeaponDefs[weaponDefID].customParams.smart_trajectory_checker then
+					unitDefData[unitDefID] = unitDefData[unitDefID] or {}
+					unitDefData[unitDefID].trajectoryCheckWeapon = weaponNumber
+				end
+			end
+		end
+	end
+end
+
+
+function gadget:UnitCreated(unitID, unitDefID)
+	if unitDefData[unitDefID] then
+		smartUnits[unitID] = {
+			unitDefID = unitDefID,
+			setStateScriptID = Spring.GetCOBScriptID(unitID, "SetAimingState"),
+			aggroBias = 0,
+			failedShotFrame = 0,
+			errorTallyMultiplier = 0,
+		}
+		spCallCOBScript(unitID, smartUnits[unitID].setStateScriptID, 0, unitDefData[unitDefID].priorityWeapon)
+	end
+end
+
+local function failureToFireCheck(attackerID, data, defData)
+	if not data.suspendErrorUntilFrame then return false end
+
+	if data.failedShotFrame < gameFrame - defData.failedToFireFrameThreshold then
+		data.failedShotFrame = mathMax(
+			spGetUnitWeaponState(attackerID, defData.priorityWeapon, 'reloadFrame'),
+			spGetUnitWeaponState(attackerID, defData.backupWeapon, 'reloadFrame')
+		)
+	end
+
+	if data.failedShotFrame < gameFrame - defData.failedToFireFrameThreshold and
+		gameFrame > data.suspendErrorUntilFrame then
+		return true
+	else
+		return false
+	end
+end
+
+
+local function updateAimingState(attackerID)
+	local data = smartUnits[attackerID]
+	local defData = unitDefData[data.unitDefID]
+
+	local priorityTargetType, priorityIsUserTarget, priorityTarget = spGetUnitWeaponTarget(attackerID, defData.priorityWeapon)
+	local backupIsUserTarget, backupTarget = select(2, spGetUnitWeaponTarget(attackerID, defData.backupWeapon))
```

Comment:
`Spring.GetUnitWeaponTarget` can have four distinct sets of returned parameters, denoted by the value of the first returned parameter. This code is assuming the return type will always be type 2, targeting a map position.

Reply:
Incorrect. 
There is an if statement that detects which return value arrangement is being provided by the function. It handles the data appropriately. 

I select two because I only need those variables from two onward

Reply:
It's checking the target type against `PRIORITY_AIMINGSTATE` and `BACKUP_AIMINGSTATE`. These are not semantically related, even if the enum values happen to be the same.

Reply:
Ahh, yes. I see it now. Fixed.